spec/extra_fields_spec.rb

```diff
@@ -1,99 +1,82 @@
+    json['data'][0]['attributes']
   end
 
   it 'does not include extra fields when not requested' do
-    expect(json['data'][0]['attributes'].keys).to match_array(%w(first_name last_name))
+    render
+    expect(attributes.keys).to match_array(%w(first_name last_name age))
   end
 
   it 'includes the extra fields in the response when requested' do
-    params[:extra_fields] = { authors: 'net_worth' }
-    expect(json['data'][0]['attributes'].keys).to match_array(%w(first_name last_name net_worth))
+    params[:extra_fields] = { employees: 'stack_ranking' }
+    render
+    expect(attributes.keys)
+      .to match_array(%w(first_name last_name age stack_ranking))
   end
 
-  it 'alters the scope based on the supplied block' do
-    params[:extra_fields] = { authors: 'net_worth' }
-    expect(json['data'][0]['attributes'].keys).to match_array(%w(first_name last_name net_worth))
+  context 'when altering scope based on extra attrs' do
+    before do
+      resource.class.class_eval do
+        extra_field :net_worth do |scope|
+          scope[:foo] = 'bar'
+          scope
+        end
+      end
+    end
+
+    it 'modifies the scope' do
+      params[:extra_fields] = { employees: 'net_worth' }
+      expect(PORO::DB).to receive(:all).with(hash_including(foo: 'bar'))
+      render
+    end
   end
 
-  context 'when accessing runtime context' do
+  context 'when acessing runtime context' do
     before do
-      params[:extra_fields] = { authors: 'runtime_id' }
+      params[:extra_fields] = { employees: 'runtime_id' }
     end
 
     it 'works' do
-      expect(scope_object).to receive(:runtime_id=).with(789)
       ctx = double(runtime_id: 789).as_null_object
       resource.with_context ctx do
-        attrs = json['data'][0]['attributes']
-        expect(attrs['runtime_id']).to eq(789)
+        render
+        expect(attributes['runtime_id']).to eq(789)
       end
     end
   end
 
-  context 'when extra field is requested but guarded' do
+  context 'when extra field is guarded' do
     before do
-      params[:extra_fields] = { authors: 'net_worth' }
+      params[:extra_fields] = { employees: 'admin_stack_ranking' }
     end
 
-    it 'does not include the extra field in the response' do
-      ctx = double(allow_net_worth?: false).as_null_object
-      resource.with_context ctx do
-        expect(json['data'][0]['attributes'].keys).to match_array(%w(first_name last_name))
+    context 'and the guard passes' do
+      it 'renders the field' do
+        ctx = double(current_user: 'admin').as_null_object
+        resource.with_context ctx do
+          render
+          expect(attributes.keys).to include('admin_stack_ranking')
+        end
+      end
+    end
+
+    context 'and the guard fails' do
+      it 'does not render the field' do
+        ctx = double(current_user: 'foo').as_null_object
+        resource.with_context ctx do
+          render
+          expect(attributes.keys).to_not include('admin_stack_ranking')
+        end
```

Comment:
Should also include a test for when a non-existent extra field is requested. I go back and forth on whether those should raise an error or should simply ignore the request the same way the field is dropped if the guard fails in the above test.  If we're assuming that anyone who can access the API knows all possible fields (even if they can't access them), then the former is preferred.  If the presence of a field can be presumed as secret without the right context, then we should do the latter. I think with our schema and all that upcoming the former is probably correct?

Reply:
Agreed. My tentative thought is we do an upcoming PR where we randomly cause failures in the sample app and ensure helpful errors are raised. Making a note of this one specifically.